curvefs/test/metaserver/copyset/concurrent_apply_queue_test.cpp

```diff
@@ -0,0 +1,222 @@
+    }
+
+    {
+        // 5. double init
+        ApplyOption opt(1, 1, 1, 1);
+        // init with vaild params
+        ASSERT_TRUE(concurrentapply.Init(opt));
+        // re-init
+        ASSERT_TRUE(concurrentapply.Init(opt));
+    }
+
+    concurrentapply.Stop();
+    // re-stop
+    concurrentapply.Stop();
+}
+
+void init_read_write_type_list(std::vector<OperatorType> *read_type_list,
+                               std::vector<OperatorType> *write_type_list) {
+    *read_type_list = {
+        OperatorType::GetDentry,
+        OperatorType::ListDentry,
+        OperatorType::GetInode,
+        OperatorType::BatchGetInodeAttr,
+        OperatorType::BatchGetXAttr,
+        OperatorType::GetVolumeExtent
+    };
+
+    auto is_read = [&read_type_list](OperatorType type) -> bool {
+        for (auto &read_type : *read_type_list) {
+            if (type == read_type) {
+                return true;
+            }
+        }
+        return false;
+    };
+
+    for (uint32_t i = 0; i < static_cast<uint32_t>(OperatorType::OperatorTypeMax); ++i) {  // NOLINT
+        if (!is_read(static_cast<OperatorType>(i))) {
+            write_type_list->push_back(static_cast<OperatorType>(i));
+        }
+    }
+}
+
+OperatorType get_random_type(const std::vector<OperatorType> &type_list) {
+    return type_list[rand() % type_list.size()];
+}
+
+TEST(ApplyQueue, RunTest) {
+    std::vector<OperatorType> read_type_list;
+    std::vector<OperatorType> write_type_list;
+    init_read_write_type_list(&read_type_list, &write_type_list);
+
+    auto read_type  = get_random_type(read_type_list);
+    auto write_type = get_random_type(write_type_list);
+
+    ApplyQueue concurrentapply;
+
+    int testw = 0;
+    auto wtask = [&testw]() {
+        std::this_thread::sleep_for(std::chrono::milliseconds(500));
+        testw++;
+    };
+
+    int testr = 0;
+    auto rtask = [&testr]() {
+        testr++;
+    };
+
+    // push write and read tasks
+    ApplyOption opt(1, 1, 1, 1);
+
+    ASSERT_TRUE(concurrentapply.Init(opt));
+
+    ASSERT_TRUE(concurrentapply.Push(1, read_type, rtask));
+    ASSERT_TRUE(concurrentapply.Push(1, write_type, wtask));
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(200));
+    ASSERT_EQ(1, testr);
+    ASSERT_EQ(0, testw);
+    concurrentapply.Flush();
+    ASSERT_EQ(1, testw);
+
+    concurrentapply.Stop();
+}
+
+TEST(ApplyQueue, FlushTest) {
+    std::vector<OperatorType> read_type_list;
+    std::vector<OperatorType> write_type_list;
+    init_read_write_type_list(&read_type_list, &write_type_list);
+
+    ApplyQueue concurrentapply;
+    ApplyOption opt(2, 5000, 1, 1);
+    ASSERT_TRUE(concurrentapply.Init(opt));
+
+    std::atomic<uint32_t> testnum(0);
+    auto task = [&testnum]() {
+        testnum.fetch_add(1);
+    };
+
+    for (int i = 0; i < 5000; i++) {
+        auto write_type = get_random_type(write_type_list);
+        concurrentapply.Push(i, write_type, task);
+    }
+
+    ASSERT_LT(testnum, 5000);
```

Comment:
Would it be better to use `ASSERT_LE`? Because `testnum` can be 5000 theoretically.

Reply:
done